### PR TITLE
[C4GT] - Support for Optional Material 

### PR DESCRIPTION
**Pull request by Saurabhbalke**
## overwiew 

- Contents marked as optional should be excluded from the course progress calculation.

-  This PR fixes [Support for Optional Material in Course #8664 ](https://github.com/Sunbird-Ed/SunbirdEd-portal/issues/8664)

### Code Requirement 

- [x] The Optional material shouldn’t contribute to the users' progress % of the trackable collection and hence doesn’t contribute in the completion certificate criteria. For ex., if a course (which is a trackable collection) has 3 resources - 2 mandatory and 1 optional. If the batch which is offering this course has a certificate issue criteria = ‘issue certs to all users whose course progress is 100%’ then the certificate should be issued as soon as the user completes the 2 mandatory resources.

- [x]  The optional material shouldn’t contribute to the final score calculation of the collection and hence doesn’t contribute in the merit certificate criteria. For ex., if a course (which is a trackable collection) has 3 resources - 2 mandatory and 1 optional. The optional resource is a course assessment. In this scenario, the certificate issue criteria should not show the option to issue certs based on score. Any batches which are offering this course should only be able to issue completion certificates. In the course ToC page, user should be able to view their latest attempt score. In the course progress exhaust as well the assessment score details should be shown.

- [x]  Additional material can be a content within a module or all of them can be bundled together as a module.

---

### Proof that changes are correct
[cinnamon-2023-09-06T171015+0530.webm](https://github.com/Sunbird-Ed/SunbirdEd-portal/assets/86031203/d6caaae5-c89e-47d2-bb6d-9dd5239216e3)

---

### Type of change

Please choose appropriate options.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Enhancement (additive changes to improve performance)
- [ ] This change requires a documentation update

### Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by Saurabhbalke** [comment]
@rajeevsathish sir, @princegupta1131 sir PTAL!!

File: src/app/client/src/app/modules/learn/components/course-consumption/assessment-player/assessment-player.component.ts [hunk 1]
```diff
@@ -541,10 +543,11 @@ export class AssessmentPlayerComponent implements OnInit, OnDestroy, ComponentCa
           unit.isUnitConsumed = consumedContent.length === 1;
           unit.progress = consumedContent.length ? 100 : 0;
           unit.isUnitConsumptionStart = Boolean(consumedContent.length);
+          if(unit?.relationalMetadata?.optional=== false) {
```
Comment (princegupta1131) [inline src/app/client/src/app/modules/learn/components/course-consumption/assessment-player/assessment-player.component.ts:546]: @Saurabhbalke please describe the code in a comment

Comment (Saurabhbalke) [reply src/app/client/src/app/modules/learn/components/course-consumption/assessment-player/assessment-player.component.ts]: hii @princegupta1131  sir,
added comments on the changes, PTAL!
thanks

File: src/app/client/src/app/modules/learn/components/course-consumption/assessment-player/assessment-player.component.ts [hunk 2]
```diff
@@ -512,7 +512,8 @@ export class AssessmentPlayerComponent implements OnInit, OnDestroy, ComponentCa
 
           /* istanbul ignore else */
           if (_.get(unit, 'children.length')) {
-            flattenDeepContents = this.courseConsumptionService.flattenDeep(unit.children).filter(item => item.mimeType !== 'application/vnd.ekstep.content-collection');
+            // exclude optional material in content list of course module
```
Comment (princegupta1131) [inline src/app/client/src/app/modules/learn/components/course-consumption/assessment-player/assessment-player.component.ts:515]: @Saurabhbalke please user proper indentation remove the unsued whitespace

File: src/app/client/src/app/modules/learn/components/course-consumption/assessment-player/assessment-player.component.ts [hunk 3]
```diff
@@ -512,7 +512,8 @@ export class AssessmentPlayerComponent implements OnInit, OnDestroy, ComponentCa
...
+            flattenDeepContents = this.courseConsumptionService.flattenDeep(unit.children).filter(item => item.mimeType !== 'application/vnd.ekstep.content-collection' && item?.relationalMetadata?.optional === false);
```
Comment (princegupta1131) [inline src/app/client/src/app/modules/learn/components/course-consumption/assessment-player/assessment-player.component.ts:516]: @Saurabhbalke can you please try this && !(item?.relationalMetadata?.optional)

File: src/app/client/src/app/modules/learn/components/course-consumption/course-player/course-player.component.ts [hunk 4]
```diff
@@ -595,6 +604,10 @@ export class CoursePlayerComponent implements OnInit, OnDestroy {
           unit.isUnitConsumed = consumedContent.length === 1;
           unit.progress = consumedContent.length ? 100 : 0;
           unit.isUnitConsumptionStart = Boolean(consumedContent.length);
+          if(unit?.relationalMetadata?.optional=== false) {
```
Comment (princegupta1131) [inline src/app/client/src/app/modules/learn/components/course-consumption/course-player/course-player.component.ts:607]: @Saurabhbalke please try to make it more concise using this if (!unit?.relationalMetadata?.optional)

File: src/app/client/src/app/modules/learn/components/course-consumption/course-player/course-player.component.ts [hunk 5]
```diff
@@ -566,7 +571,7 @@ export class CoursePlayerComponent implements OnInit, OnDestroy {
 
           /* istanbul ignore else */
           if (_.get(unit, 'children.length')) {
-            flattenDeepContents = this.courseConsumptionService.flattenDeep(unit.children).filter(item => item.mimeType !== 'application/vnd.ekstep.content-collection' && item.mimeType !== 'application/vnd.sunbird.question');
+            flattenDeepContents = this.courseConsumptionService.flattenDeep(unit.children).filter(item => item.mimeType !== 'application/vnd.ekstep.content-collection' && item.mimeType !== 'application/vnd.sunbird.question' && item?.relationalMetadata?.optional === false);
```
Comment (princegupta1131) [inline src/app/client/src/app/modules/learn/components/course-consumption/course-player/course-player.component.ts:574]: @Saurabhbalke please try to make it more concise using this && !(item?.relationalMetadata?.optional)

File: src/app/client/src/app/modules/learn/components/course-consumption/course-player/course-player.component.ts (hunk 4 above)
Comment (Saurabhbalke) [reply src/app/client/src/app/modules/learn/components/course-consumption/course-player/course-player.component.ts]: done

**Review by princegupta1131 — APPROVED** [review]

**Pull request closed without merging**